retry.go

```diff
@@ -25,60 +26,95 @@ func init() {
 }
 
+// requestGUIDKey is attached to every request against Snowflake
 const requestGUIDKey string = "request_guid"
 
-// Format of "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
-const uuidLen int = 36
+// retryCounterKey is attached to query-request from the second time
+const retryCounterKey string = "retryCounter"
+
+// requestIDKey is attached to all requests to Snowflake
+const requestIDKey string = "requestId"
 
 // This class takes in an url during construction and replace the
 // value of request_guid every time the replace() is called
 // When the url does not contain request_guid, just return the original
 // url
 type requestGUIDReplacerI interface {
 	// replace the url with new ID
-	replace() string
+	replace() *url.URL
 }
 
 // Make requestGUIDReplacer given a url string
-func makeRequestGUIDReplacer(url string) requestGUIDReplacerI {
-	startIndex := strings.Index(url, requestGUIDKey)
-	if startIndex == -1 {
-		return &transientReplacer{url}
+func makeRequestGUIDReplacer(urlPtr *url.URL) requestGUIDReplacerI {
+	_, err := url.ParseQuery(urlPtr.RawQuery)
+	if err != nil {
+		return &transientReplacer{urlPtr}
 	}
-	replacer := &requestGUIDReplacer{}
-	startIndex += len(requestGUIDKey) + 1
-	replacer.prefix = url[:startIndex]
-
-	startIndex += uuidLen
-	replacer.suffix = url[startIndex:]
-	return replacer
+	return &requestGUIDReplacer{urlPtr}
 }
 
 // this replacer does nothing but replace the url
 type transientReplacer struct {
-	url string
+	urlPtr *url.URL
 }
 
-func (replacer *transientReplacer) replace() string {
-	return replacer.url
+func (replacer *transientReplacer) replace() *url.URL {
+	return replacer.urlPtr
 }
 
 /*
 requestGUIDReplacer is a one-shot object that is created out of the retry loop and
 called with replace to change the retry_guid's value upon every retry
 */
 type requestGUIDReplacer struct {
-	// cached prefix and suffix to avoid parsing same url again
-	prefix string
-	suffix string
+	urlPtr *url.URL
 }
 
 /**
 This function would replace they value of the requestGUIDKey in a url with a newly
 generated uuid
 */
-func (replacer *requestGUIDReplacer) replace() string {
-	return replacer.prefix + uuid.New().String() + replacer.suffix
+func (replacer *requestGUIDReplacer) replace() *url.URL {
+	vs, err := url.ParseQuery(replacer.urlPtr.RawQuery)
+	if err != nil {
+		return replacer.urlPtr
+	}
+	if len(vs.Get(requestGUIDKey)) == 0 {
+		return replacer.urlPtr
+	}
+	vs.Del(requestGUIDKey)
+	vs.Add(requestGUIDKey, uuid.New().String())
+	replacer.urlPtr.RawQuery = vs.Encode()
+	return replacer.urlPtr
+}
+
+type retryUpdater interface {
+	replaceOrAdd(retry int) *url.URL
+}
+
+type retryUpdate struct {
+	TargetURL *url.URL
+}
+
+func (r *retryUpdate) replaceOrAdd(retry int) *url.URL {
+	if !strings.HasPrefix(r.TargetURL.Path, queryRequestPath) {
+		return r.TargetURL
+	}
+
+	vs, err := url.ParseQuery(r.TargetURL.RawQuery)
+	if err != nil {
+		return r.TargetURL
+	}
```

Comment:
I would probably either put these logic in the constructor and have a another struct named transienUpdate as consistent with GUIDReplacer and get slightly better performance in the case there are lot of retries, or just implement this as a simple function with no OOP at all.